File: libs/sdk-ts/src/stream.ts
// Copyright 2026 GAIA Contributors
//
// Licensed under the MIT License.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * This module implements the WebSocket stream listener for the GAIA TypeScript SDK.
 * It provides a reactive interface for observing the GAIA 10-phase control loop
 * in real-time as events are emitted by the Kernel.
 */
import { WebSocket } from 'ws';
import { Event } from './types';
import { EventEmitter } from 'events';

/**
 * GaiaStream connects to the GAIA Kernel's real-time event stream.
 * It allows developers to react to step completions, planning updates, and failures
 * without polling the REST API, ensuring low-latency observability.
 */
export class GaiaStream extends EventEmitter {
  private ws: WebSocket | null = null;
  private url: string;

  constructor(baseURL: string = 'ws://localhost:8080') {
    super();
    // Convert http/https to ws/wss if needed
    this.url = baseURL.replace(/^http/, 'ws') + '/api/v1/stream';
  }

  /**
   * connect opens the persistent WebSocket connection.
   */
  connect(): void {
    this.ws = new WebSocket(this.url);

    this.ws.on('open', () => {
      this.emit('connected');
    });

    this.ws.on('message', (data: string) => {
      try {
        const event: Event = JSON.parse(data);
        this.emit('event', event);
        this.emit(event.type, event);
      } catch (err) {
        this.emit('error', new Error('Failed to parse event: ' + err));
      }
    });

    this.ws.on('close', () => {
      this.emit('disconnected');
    });

    this.ws.on('error', (err) => {
      this.emit('error', err);
    });
  }

  /**
   * disconnect closes the WebSocket connection gracefully.
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}

File: libs/sdk-py/gaia/stream.py
# Copyright 2026 GAIA Contributors
#
# Licensed under the MIT License.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
This module implements the WebSocket stream listener for the GAIA Python SDK.
It provides an asynchronous, event-driven interface for observing Kernel events.
"""

import json
import asyncio
import websockets
from typing import Callable, Dict, List, Any

class GaiaStream:
    """
    GaiaStream connects to the GAIA Kernel's real-time event stream.
    It allows developers to register callbacks for specific event types.
    """
    def __init__(self, base_url: str = "http://localhost:8080"):
        self.url = base_url.replace("http", "ws") + "/api/v1/stream"
        self.callbacks: Dict[str, List[Callable]] = {}
        self.ws = None

    def on(self, event_type: str, callback: Callable):
        """
        Registers a callback for a specific event type.
        """
        if event_type not in self.callbacks:
            self.callbacks[event_type] = []
        self.callbacks[event_type].append(callback)

    async def listen(self):
        """
        Connects to the stream and begins processing messages.
        """
        async with websockets.connect(self.url) as ws:
            self.ws = ws
            async for message in ws:
                try:
                    event = json.loads(message)
                    event_type = event.get("type")
                    
                    # Trigger specific callbacks
                    if event_type in self.callbacks:
                        for cb in self.callbacks[event_type]:
                            if asyncio.iscoroutinefunction(cb):
                                await cb(event)
                            else:
                                cb(event)
                    
                    # Trigger catch-all callbacks
                    if "*" in self.callbacks:
                        for cb in self.callbacks["*"]:
                            if asyncio.iscoroutinefunction(cb):
                                await cb(event)
                            else:
                                cb(event)
                                
                except Exception as e:
                    print(f"Error processing stream event: {e}")

    async def stop(self):
        if self.ws:
            await self.ws.close()
